autopilot/agent.go

```diff
@@ -24,14 +24,15 @@ type Config struct {
 	// accidentally attempt to open a channel with ourselves.
 	Self *btcec.PublicKey
 
-	// Heuristic is an attachment heuristic which will govern to whom we
-	// open channels to, and also what those channels look like in terms of
-	// desired capacity. The Heuristic will take into account the current
-	// state of the graph, our set of open channels, and the amount of
-	// available funds when determining how channels are to be opened.
-	// Additionally, a heuristic make also factor in extra-graph
-	// information in order to make more pertinent recommendations.
-	Heuristic AttachmentHeuristic
+	// Heuristics is a slice of attachment heuristics which will govern to
+	// whom we open channels to. The Heuristics will take into account the
+	// current state of the graph, our set of open channels, and desired
+	// channel size when determining how channels are to be opened.
+	// Additionally, a heuristic might also factor in extra-graph
+	// information in order to make more pertinent recommendations. The
+	// weight given to each heuristic is a given each heuristic's weight,
+	// which must sum to 1.0 for the slice.
+	Heuristics []*WeightedHeuristic
```

Comment:
Alternatively, in order to keep the logic at the agent simple, we can create a new composite node score heuristic. So from the PoV of the agent, it queries a _single_ heuristic which then implements the weighted heuristic logic underneath. In a sense, it's just an additional layer of indirection that keeps the agent simple and the bulk of the logic within the composite node score heuristic.

Reply:
By abstracting the selection logic, we can eventually experiment with other things like the original "information gain model", the self-aware model (takes into account how its own score will be modified by accepting/making a channel), and so on.

Reply:
I went the composite heuristic way first, but it turned out to be more complex than necessary when I wanted to query each individual sub-heuristic for the dry run.

But I do agree that it would be nice to keep the agent as simple as possible, and it would possibly also make it easier to experiment with more advanced combinations than just a weighted average.

Will see if I can combine the two approaches, still keeping the dryrun easy to implement.

Reply:
Done!